### PR TITLE
#39: Add support for union types (closes #39)

### LINKED ISSUE #39: Add support for union types
Metarpheus should support union types such as:

```scala
sealed abstract trait UnionExample
object UnionExample {
  case class First(arg: String) extends UnionExample
  case class Second(arg: String, anotherArg: String) extends UnionExample
}
```

Also opened https://github.com/buildo/metarpheus-io-ts/issues/79 to add the generation to `metarpheus-io-ts`

### PR DESCRIPTION
**Pull request by tpetrucciani**
Closes #39

Adds some support for tagged unions in metarpheus models.

- A `sealed trait` followed by an object containing only case objects or case classes gets recognized as a tagged union
- In the model we record for each member of the object similar information as for case classes
- Note that, previously, the same pattern with only case objects was recognized as an enumeration; now we only get enumerations when using the annotations from enumero

Also, I've added a parameter `error` to routes expressing the error type, e.g. `E` if the return type is `Future[Either[E, T]]`. It's optional because metarpheus also accepts return types like `Future[T]`.

Note that:

- currently, we get duplicate elements in the model to represent case classes in tagged unions, as discussed [here](https://github.com/buildo/retro/issues/39#issuecomment-504374343); there are probably multiple ways to fix it which could be discussed
- I'm ignoring type parameters in tagged unions

## Test Plan

I've added to the tests an example of tagged union containing both case objects and case classes, checking the obtained model and also that `stripUnusedModels` is correctly updated to account for error types in routes and parameters in tagged unions.

### CODE REVIEW

File: metarpheus/core/src/main/scala/io.buildo.metarpheus/core/extractors/model.scala [hunk 1]
```diff
@@ -48,41 +48,47 @@ package object model {
...
+      case c =>
+        c.children
+          .sliding(2)
+          .filter {
+            case (t: Defn.Trait) :: (o: Defn.Object) :: Nil =>
+              t.mods.collectFirst {
+                case _: Mod.Sealed => ()
+              }.isDefined &&
+                o.templ.stats.forall {
+                  case c: Defn.Class if c.mods.collectFirst {
```
Comment (gabro) [inline metarpheus/core/src/main/scala/io.buildo.metarpheus/core/extractors/model.scala]: You can import `import scala.meta.contrib._` and then use `hasMod` like:

```scala
case c: Defn.Class if c.hasMod(Mod.Case) =>
```

File: metarpheus/core/src/main/scala/io.buildo.metarpheus/core/intermediate/intermediate.scala [hunk 2]
```diff
@@ -28,12 +28,24 @@ object CaseEnum {
   case class Member(name: String, desc: Option[String])
 }
 
+case class TaggedUnion(name: String, values: List[TaggedUnion.Member], desc: Option[String])
+    extends Model
+object TaggedUnion {
+  case class Member(
+    name: String,
+    params: List[MemberParam],
```
Comment (gabro) [inline metarpheus/core/src/main/scala/io.buildo.metarpheus/core/intermediate/intermediate.scala]: I think we can just reuse `CaseClass.Member` here

File: metarpheus/core/src/main/scala/io.buildo.metarpheus/core/intermediate/intermediate.scala [hunk 3]
```diff
@@ -28,12 +28,24 @@ object CaseEnum {
...
+    desc: Option[String],
+    isValueClass: Boolean = false,
+  )
+  case class MemberParam(name: String, tpe: Type, desc: Option[String])
```
Comment (gabro) [inline metarpheus/core/src/main/scala/io.buildo.metarpheus/core/intermediate/intermediate.scala]: see above, this is probably redundant

File: metarpheus/core/src/main/scala/io.buildo.metarpheus/core/extractors/controller.scala [hunk 4]
```diff
@@ -28,53 +28,53 @@ package object controller {
...
+                             |The return type can be of two types:
+                             | - F[E[_, Result]]
+                             | - F[Result],
         """.stripMargin)
-      }
+    }
+
+  private[this] def extractErrorType(m: Decl.Def): Option[intermediate.Type] =
+    m.decltpe.collect {
+      case Type.Apply(Type.Name(_), Seq(Type.Apply(Type.Name(_), Seq(tpe, _)))) =>
```
[the remark this answers is not in the document]
Comment (tpetrucciani) [reply metarpheus/core/src/main/scala/io.buildo.metarpheus/core/extractors/controller.scala]: I'd skip using these until we figure out about the unused variables warning

**Pull request merged**